### PR TITLE
[FIX] web: fix kanban useSortable screen width limit + [IMP] web: add edge scrolling behavior to useSortable

### PR DESCRIPTION
**Pull request by SimonGenin**
# [FIX] web: fix kanban useSortable screen width limit
## Reproduce:
Go on a kanban view, with enough columns so it goes off the screens.
Scroll on the right, try to drag anything beyond the original (before
scrolling) limit of the screen. The cards are stuck floating at that
limit. Note it didn't prevent the drop to work properly where the mouse
was.

## Issue:
The useSortable hook gets an owl ref, which is used as a limit to where
things can be dragged and dropped. Originaly and logically, it was the
current component ref, the renderer. But for some unknown reason, this
space was not getting wider than the initial screen size. It may be a
problem with the flexbox layout.

## Tried solutions:
- A first solution was to add the overflow-x-scroll css property to the
renderer. It works, but now the horizontal scroll bar is no longer
always visible. You have to scroll all the way down to see it. UX wise
this is not acceptable.

- We could remove the clamp mecanism inside useSortable to let the drag
and drop motion go anywhere. This works, but decrease the quality of the
interaction. It looks cheap and not polished.

## Final Solution
Finally, we decided to pass a ref to the o_content div from the layout
component down to the kanban renderer. If the useSortable is given this
ref, it works as intended. While it may seem too much, it is not
unreasonable to propose a reference to the most parent element of the
view contents to the renderer.

# [IMP] web: add edge scrolling behavior to useSortable
In house implementation of automatic scoll when dragging an element to
the edges of its container.

Enterprise PR: https://github.com/odoo/enterprise/pull/32478

### CODE REVIEW

**Comment by SimonGenin** [comment]
Note about testing:
After hours of trying, it was not possible to make a test for this. The tried test consisted in checking if a card had been farther than some column, proving that the card had gone beyond the screen. 
But, even after a few tweaks inside the useSortable, the final result would show that the final position of the card would always be the edge of the screen (even when the bug is no longer there obvisouly). Probably because it is the last place when it is renderered or something.

File: addons/web/static/src/core/utils/draggable_hook_builder.js [hunk 1]
```diff
@@ -275,6 +298,51 @@ export function makeDraggableHook(hookParams = {}) {
                 }
             };
 
+            /**
+             * Applies scroll to the container if the current element is near
+             * the edge of the container.
+             */
+            const handleEdgeScrolling = (deltaTime) => {
+                if (!state.dragging) {
```
Comment (SimonGenin) [inline addons/web/static/src/core/utils/draggable_hook_builder.js:306]: it that necessary ?

File: addons/web/static/src/core/utils/sortable.js [hunk 2]
```diff
@@ -66,16 +66,15 @@ export const useSortable = makeDraggableHook({
         groups: ["string", "function"],
         connectGroups: ["boolean", "function"],
     },
+    defaultParams: {
+        connectGroups: false,
+        currentGroup: null,
+        edgeScrolling: { speed: 20, threshold: 40 },
```
Comment (SimonGenin) [inline addons/web/static/src/core/utils/sortable.js:72]: With the treshold, after trying it on the runbot, I think we should increase it, perhaps even double it.

Comment (SimonGenin) [reply addons/web/static/src/core/utils/sortable.js:72]: or 60

File: addons/web/static/tests/views/kanban_view_tests.js [hunk 3]
```diff
@@ -12077,4 +12077,90 @@ QUnit.module("Views", (hooks) => {
...
+        let animationFrameDef = makeDeferred();
+        let timeStamp = 0;
+
+        patchWithCleanup(browser, {
+            async requestAnimationFrame(handler) {
+                await animationFrameDef;
+                handler(timeStamp);
+            },
+            performance: { now: () => timeStamp },
+        });
```
Comment (SimonGenin) [inline addons/web/static/tests/views/kanban_view_tests.js:12098]: damn, nice work 👍

**Comment by Iucapad** [comment]
Nice addition :) Good job!

File: addons/web/static/tests/setup.js [hunk 4]
```diff
@@ -171,8 +173,21 @@ function patchBrowserWithCleanup() {
...
+                    if (animationFrameHandles.has(handle)) {
+                        fn(16);
+                    }
+                });
+
+                return handle;
+            },
+            cancelAnimationFrame: (handle) => {
+                animationFrameHandles.delete(handle);
+            },
```
Comment (aab-odoo) [inline addons/web/static/tests/setup.js:190]: I hope this doesn't introduce sources of non deterministic errors on runbot. Was this strictly necessary? I thought we talked about only patching requestAnimationFrame in specific tests.

**Review by aab-odoo — APPROVED** [review]
Impressive. This is too technical for a advanced review, so I rather manually tested on runbot and it seems to work pretty well. Good job :+1: 

My only concern is about the requestAnimationFrame mock, I'll r+ as soon as this is clarified.

**Comment by aab-odoo** [comment]
Thanks for the explanation @Arcasias 

robodoo r+

**Comment by aab-odoo** [comment]
robodoo rebase-ff

**Pull request closed without merging**